### PR TITLE
AP-1713 Store page history in redis

### PR DESCRIPTION
**Pull request by mpw5**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/AP-1713)

To prevent cookies getting too large, move storage of page history to redis instead of the session.

- Add a service class to write/read page history records to/from redis, based on a key based on a `page_history_id` session variable which is set when an applicant or provider starts a new session on the service.

- Amend `Backable` module and `ConfirmOfficesController` to use the new service class instead of using the session to store the page history.

- Fix specs and feature tests to ensure that a consistent session id is available throughout a test, to allow data to be be written/read correctly.

## Checklist

Before you ask people to review this PR:

- [x] Tests and rubocop should be passing: `bundle exec rake`
- [x] Github should not be reporting conflicts; you should have recently run `git rebase master`.
- [x] There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- [x] The PR description should say what you changed and why, with a link to the JIRA story.
- [x] You should have looked at the diff against master and ensured that nothing unexpected is included in your changes.
- [x] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

**Review by stephenrichards — COMMENTED** [review]
I'm beginning to think that it would be better to generate a unique id and store that in the session rather than use the session id itself.  We just don't know enough about whether the session id changes in real life or not, as well as in the specs.  Also, you might want to consider using a separate database (redis has 16 of them numbered 0 to 15, and uses 0 as the default).

Neither of these comments are show stoppers - it just might be something worth considering.

File: app/controllers/concerns/backable.rb [hunk 1]
```diff
@@ -17,6 +17,7 @@ def skip_back_history_for(*actions)
 
     # Adding ?back=true to the path of the previous page to indicate that we are moving backward in the page history
     def back_path
+      # binding.pry
       return unless page_history.size > 1
```
[the remark this answers is not in the document]
Comment (mpw5) [reply app/controllers/concerns/backable.rb]: I can't see that in my version of this PR 🤔 Maybe you need to refresh to see the latest version?

[the remark this answers is not in the document]
Comment (mpw5) [reply app/controllers/concerns/backable.rb]: I tend to agree that another id might be a better option. The session_id seems to work from a user's perspective (I've been back and forth through the application several times and it behaves as expected) but fudging the tests like this doesn't feel right. I'll look at adding another id to the session that we can use, although I suspect I might still run into issues in the specs, as it's relying on a value in the session and it seems like we can't guarantee that being consisten in request specs.

I'm already using a separate redis database for this, it's configured in the url that is called - eg 'redis://localhost:6379/1' (although I've just noticed a mistake that I'll fix now!)

**Review by stephenrichards — APPROVED** [review]

**Review by colinbruce — COMMENTED** [review]
Does this mean that redis now _must_ be running?  If so, can you add that to the readme in the 
[run-the-application-server](https://github.com/ministryofjustice/laa-apply-for-legal-aid#run-the-application-server) section?

**Comment by mpw5** [comment]
Good point @colinbruce. Yes, it does. I'll update the readme.

**Comment by colinbruce** [comment]
> Good point @colinbruce. Yes, it does. I'll update the readme.

It's only because @willc-work has been updating it and, while I was reviewing his work, I thought of this from yesterday! 🎉

**Pull request merged**